### ISSUE TITLE
False error TS4094: ... exported class expression may not be private or protected.

### ISSUE THREAD

**Issue by a-student** [Microsoft/TypeScript#30355]
**TypeScript Version:**  3.3.3333

**Code**

_src/test.ts:_
```ts
export function test () {
    return class {
        private privateMember () {
        }
    };
}
```

_tsconfig.json:_
```json
{
    "compilerOptions": {
        "target": "es5",
        "module": "commonjs",
        "declaration": true
    },
    "include": [
        "src/**/*"
    ],
    "exclude": [
        "node_modules"
    ]
}
```

**Expected behavior:**
No compilation error.

**Actual behavior:**
Compiler prints error:
$ tsc
src/test.ts(1,17): error TS4094: Property 'privateMember' of exported class expression may not be private or protected.

**Playground Link:** not possible to provide.

**Workaround**

Declare the return type explicitly:

```ts
export function test (): new() => Object {
    return class {
        private privateMember () {
        }
    };
}
```

**Comment by RyanCavanaugh** [comment]
The error message didn't accidently write itself... exported anonymous classes can't have private or protected members if declaration emit is enabled, because there's no way to represent that in a .d.ts file.

**Comment by a-student** [comment]
It is possible to use private in declaration files. In my project, for example, I found a lot of them in _node_modules_ folder:
_tslint/lib/formatters/checkstyleFormatter.d.ts:_
```ts
export declare class Formatter extends AbstractFormatter {
    static metadata: IFormatterMetadata;
    format(failures: RuleFailure[]): string;
    private escapeXml; <- HERE
}
```

**Comment by taoqf** [comment]
I think this is an issue, too.

**Comment by AlexLeung** [comment]
Perhaps we should extends d.ts to handle this case then. Ideally the declarations file and the ts source file would be 100% compatible.

**Comment by cdata** [comment]
It is currently not possible to generate declaration files when using a class mixin pattern, even if that pattern would otherwise compile and type-check just fine. It would be nice to have support for this 🙏

**Comment by trusktr** [comment]
Shouldn't this issue be open?

**Comment by trusktr** [comment]
In particular, it'd be nice to be able to do something like

```ts
function foo<T>(arg: T): SomeType<T> { ... }

type Bar = {num: number}

type FooBar = ReturnType<typeof foo<Bar>>
```

I get what you're saying, that `foo<Bar>` is already a type, so typeof doesn't make sense. But I think you get what's missing.

This is valid:

```ts
type FooBar = ReturnType<typeof foo>
```

But in that example, TypeScript sets the type of `T` to `unknown`, so the type of FooBar is `SomeType<T>`.

Basically, it just intuitively seems like there should be some way to pass a generic arg there, but we can't, and TypeScript automatically sticks `unknown` into it. [Example on playground](http://www.typescriptlang.org/play/#code/C4TwDgpgBAyg9gWwgFXBAPMgfFAvFAQQCciBDETLAWAChaATCAYwBtSjoAzAVwDsngASzi8onOHEoAKdgHMAXFGQBKRfCSpIlWrVCQoAIXZ4oAb17cEiiwgBGEIgF8dNPdABiEo0RMAlCMDcRLyaGG5wnGISOAD0MVB8ANa8cADuvADaALpAA)

**Comment by trusktr** [comment]
If a function `foo` has a generic param, then the `typeof` operator could perhaps return a generic type, and we could write:

```ts
type FooBar = ReturnType<(typeof foo)<Bar>>
```

That would be a breaking change to TypeScript though.

**Comment by trusktr** [comment]
Please see https://github.com/microsoft/TypeScript/issues/35822

**Comment by trusktr** [comment]
I re-opened this at https://github.com/microsoft/TypeScript/issues/36060

**Comment by sebastiandittrich** [comment]
Interesting fix: moving the mixin to its own file works:
```ts
type Constructor<T = {}> = new (...args: any[]) => T

export default <T extends Constructor<Interface>>(base: T) => class SomeClass extends base {
    protected something: boolean = false
}
```

While a named export throws the error:
```ts
type Constructor<T = {}> = new (...args: any[]) => T

export const mixin = <T extends Constructor>(base: T) => class SomeClass extends base {
    protected something: boolean = false
}
```

**Comment by gumT3KBZ9zMY7kzv** [comment]
here is how you can fix it, instead of
```typescript
function makeClass() {
    return class {
         private x = 0;
    }
}
```
do
```typescript
type Constructor<T> = new () => T
interface MyInterface {
    y: number;
}
function makeClass(): Constructor<MyInterface> {
    return class {
         private x = 0;
         public y = 1;
    }
}
```

**Comment by tchakabam** [comment]
> It is currently not possible to generate declaration files when using a class mixin pattern, even if that pattern would otherwise compile and type-check just fine. It would be nice to have support for this 🙏

@cdata not exactly! :) see the prior workaround the OP pointed to:

> Declare the return type explicitly:

In the case of a mixin function (which is exactly the problem I had here too), that means you need to return the union (`&`) of the 2 constructor-types which you will mixin together. It is helpful to provide therefore an interface decl outside of the mixin function which `implements` by the class that the mixin function returns.

I can post some helper code and type utilities for this if there is interest.

I agree this isn't a bug per-se. The type-declaration just expects your mixin function to return types explicitely that it actually can express i.e export in a d.ts, and in this case that will mean explicitely declaring the resulting mixin interface *outside* of the mixin function so that it can be used as a return type of it.

**Comment by Julien-R44** [comment]
Another solution can be to just use the ES2020 private fields if you ever encounter this problem. This will avoid you to define a return type, which can be annoying if your mixin is quite complex :

```ts
function makeClass() {
  return class {
       #x = 0;
       public y = 1;
  }
}
```

**Comment by leppaott** [comment]
Same thing with "constant class".

```
export const testClass = class {
  private field = {}
}
```

Interestingly `#` private class fields indeed work fine.

**Comment by Marckon** [comment]
Although use `#`to declare private member can avoid this declaration error, it will make the property not readable, plus it can not replace `protected`. **I only want some method or property not overridable in some cases**.

**Comment by zyf0330** [comment]
you can write like this
```
class C {
    private privateMember () {
    }
}
export function test ()  {
    return C;
}

export function test2 (): typeof C  {
    return class extends C {};
}
```

**Comment by ThePlenkov** [comment]
> Although use `#`to declare private member can avoid this declaration error, it will make the property not readable, plus it can not replace `protected`. **I only want some method or property not overridable in some cases**.

it would be great if ESlint rule will be already aware. I had some error and quick fix was generation of private _data + get/set data methods. And it was throwing errors. Replacing with # helped to solve it.

**Comment by ThePlenkov** [comment]
> **Interesting fix:**
> 
> Moving the mixin to its own file and doing a `default` export works:
> 
> ```ts
> type Constructor<T = {}> = new (...args: any[]) => T
> 
> export default <T extends Constructor>(base: T) => class SomeClass extends base {
>     protected something: boolean = false
> }
> ```
> 
> While a named export throws the error:
> 
> ```ts
> type Constructor<T = {}> = new (...args: any[]) => T
> 
> export const mixin = <T extends Constructor>(base: T) => class SomeClass extends base {
>     protected something: boolean = false
> }
> ```

isn't it a bug then? So it's doing exactly same, but with two files instead of one. I'm quite surprised to see that this issue is Closed while indeed it must not and should be fixed

**Comment by StiliyanKushev** [comment]
How that's still closed is beyond me. Please revise.
